internal/handlers/pg/msg_createindexes.go

```diff
@@ -75,25 +104,53 @@ func (h *Handler) MsgCreateIndexes(ctx context.Context, msg *wire.OpMsg) (*wire.
 	iter := idxArr.Iterator()
 	defer iter.Close()
 
+	duplicateChecker := map[string]struct{}{}
+
+	var isUniqueSpecified bool
+	var numIndexesBefore, numIndexesAfter int32
 	err = dbPool.InTransactionRetry(ctx, func(tx pgx.Tx) error {
+		var indexesBefore []pgdb.Index
+		indexesBefore, err = pgdb.Indexes(ctx, tx, db, collection)
+		if err == nil {
+			numIndexesBefore = int32(len(indexesBefore))
+		}
+		if errors.Is(err, pgdb.ErrTableNotExist) {
+			numIndexesBefore = 1
+			err = nil
+		}
+
+		if err != nil {
+			return lazyerrors.Error(err)
+		}
+
 		for {
-			var val any
-			_, val, err = iter.Next()
+			var key, val any
+			key, val, err = iter.Next()
 
 			switch {
 			case err == nil:
 				// do nothing
 			case errors.Is(err, iterator.ErrIteratorDone):
-				// iterator is done, no more indexes to create
+				var indexesAfter []pgdb.Index
+				indexesAfter, err = pgdb.Indexes(ctx, tx, db, collection)
+				if err != nil {
+					return lazyerrors.Error(err)
+				}
+
+				numIndexesAfter = int32(len(indexesAfter))
+
 				return nil
 			default:
 				return lazyerrors.Error(err)
 			}
 
 			indexDoc, ok := val.(*types.Document)
 			if !ok {
-				// TODO Add better validation and return proper error: https://github.com/FerretDB/FerretDB/issues/2311
-				return lazyerrors.Errorf("expected index document, got %T", val)
+				return commonerrors.NewCommandErrorMsgWithArgument(
+					commonerrors.ErrTypeMismatch,
+					fmt.Sprintf("BSON field 'createIndexes.indexes.%d' is the wrong type, expected type 'object'", key),
```

Comment:
We can use the `commonparams.AliasFromType(val)` here to make an identical error message.